src/rsz/src/Journal.cc

```diff
@@ -0,0 +1,238 @@
+
+void Journal::begin()
+{
+  debugPrint(logger_, RSZ, "journal", 1, "journal begin");
+  journal_stack_ = {};
+}
+
+void Journal::end()
+{
+  debugPrint(logger_, RSZ, "journal", 1, "journal end");
+  // TODO We need to do something with cloned instances here
+  journal_stack_ = {};
+}
+
+void Journal::swapPins(Instance* inst, LibertyPort* port1,
+                              LibertyPort* port2)
+{
+  debugPrint(logger_, RSZ, "journal", 1, "journal swap pins {} ({}->{})",
+             network()->pathName(inst), port1->name(), port2->name());
+  std::unique_ptr<Undo> element = std::make_unique<UndoPinSwap>(inst, port1, port2);
+  journal_stack_.emplace(std::move(element));
+}
+
+
+Instance *Journal::cloneInstance(LibertyCell *cell, const char *name,  Instance *original_inst,
+                                 Instance *parent,  const Point& loc)
+ {
+   Instance *clone_inst = resizer_->makeInstance(cell, name, parent, loc);
+   debugPrint(logger_, RSZ, "journal", 1, "journal clone {} ({})",
+              network()->pathName(original_inst), cell->name());
+   std::unique_ptr<Undo> element(new UndoClone(original_inst, clone_inst));
+   journal_stack_.emplace(std::move(element));
+   return clone_inst;
+}
+
+
+void Journal::instReplaceCellBefore(Instance* inst)
+{
+  LibertyCell *lib_cell = network()->libertyCell(inst);
+  debugPrint(logger_, RSZ, "journal", 1, "journal replace {} ({})",
+             network()->pathName(inst),
+             lib_cell->name());
+   std::unique_ptr<Undo> element(new UndoResize(inst, lib_cell));
+   journal_stack_.emplace(std::move(element));
+}
+
+void Journal::makeBuffer(Instance* buffer)
+{
+  debugPrint(logger_, RSZ, "journal", 1, "journal make_buffer {}",
+             network()->pathName(buffer));
+  std::unique_ptr<Undo> element(new UndoBuffer(buffer));
+  journal_stack_.emplace(std::move(element));
+}
+
+void Journal::restore()
+{
+  while (!journal_stack_.empty()) {
+    Undo *element = journal_stack_.top().get();
+    element->UndoOperation(resizer_);
+    journal_stack_.pop();
+  }
+}
+
+void Journal::reportStatistics()
+{
+  // TODO:
+  if (inserted_buffer_count_ > 0 && split_load_buffer_count_ == 0) {
+    logger_->info(RSZ, 59, "Inserted {} buffers.", inserted_buffer_count_);
+  }
+  else if (inserted_buffer_count_ > 0 && split_load_buffer_count_ > 0) {
+    logger_->info(RSZ, 87, "Inserted {} buffers, {} to split loads.",
+                  inserted_buffer_count_, split_load_buffer_count_);
+  }
+  logger_->metric("design__instance__count__setup_buffer", inserted_buffer_count_);
+  if (resize_count_ > 0) {
+    logger_->info(RSZ, 91, "Resized {} instances.", resize_count_);
+  }
+  if (swap_pin_count_ > 0) {
+    logger_->info(RSZ, 96, "Swapped pins on {} instances.", swap_pin_count_);
+  }
+  if (cloned_gate_count_ > 0) {
+    logger_->info(RSZ, 97, "Cloned {} instances.", cloned_gate_count_);
+  }
+  }
+
+sta::Network * Journal::network()
+{
+  return resizer_->network_;
+}
+
+dbSta *Journal::sta()
```

Comment:
What is this used for?

Reply:
Removed for now. Will pit it back when needed.